### DISCUSSION TITLE
PatchCore: Training only on good images and then inference on good and bad images

### DISCUSSION THREAD

**Discussion by luliuzee** [open-edge-platform/anomalib#192] [Q&A] [answered]
First of all, great work! I found this work very interesting! I have a question related to PatchCore implementation. 

When there is both training, and testing dataset (images and ground truth) available, then the PatchCore works good and I think during the validation step, it is using the testing data (GT and images) to calculate coreset subsampling. 

But imagine a realworld scenario when only good images are available to teach the anomaly detection algorithm, and no defective images are available (GT and images), then how to train the model? I've tried limit_val_batches and limit_test_batches to zero, so during training, no validation and testing are performed. But with this setup, during inference time, the model is producing this error

```
raceback (most recent call last):
  File "tools/inference.py", line 118, in <module>
    infer()
  File "tools/inference.py", line 101, in infer
    output = inference.predict(image=args.image_path, superimpose=True)
  File "/notebooks/anomalib/anomalib/deploy/inferencers/base.py", line 90, in predict
    predictions = self.forward(processed_image)
  File "/notebooks/anomalib/anomalib/deploy/inferencers/torch.py", line 118, in forward
    return self.model(image)
  File "/opt/conda/lib/python3.8/site-packages/torch/nn/modules/module.py", line 889, in _call_impl
    result = self.forward(*input, **kwargs)
  File "/notebooks/anomalib/anomalib/models/components/base/anomaly_module.py", line 75, in forward
    return self.model(batch)
  File "/opt/conda/lib/python3.8/site-packages/torch/nn/modules/module.py", line 889, in _call_impl
    result = self.forward(*input, **kwargs)
  File "/notebooks/anomalib/anomalib/models/patchcore/model.py", line 170, in forward
    patch_scores = self.nearest_neighbors(embedding=embedding, n_neighbors=9)
  File "/notebooks/anomalib/anomalib/models/patchcore/model.py", line 238, in nearest_neighbors
    distances = torch.cdist(embedding, self.memory_bank, p=2.0)  # euclidean norm
  File "/opt/conda/lib/python3.8/site-packages/torch/functional.py", line 1119, in cdist
    return _VF.cdist(x1, x2, p, None)  # type: ignore
RuntimeError: cdist only supports at least 2D tensors, X2 got: 1D
```

Any guidance will be highly appreciated!

**Comment by samet-akcay** [comment]
Hi @luliuzee , yes this a known issue. In fact, almost none of the existing anomaly detection algorithms addresses this issue. We are currently working on a solution that wouldn't require any validation or test images to find the threshold in an unsupervised manner. We have some promising results. We'll make it available here once we publish our results. 

Meanwhile, what you could do is to choose manual threshold and set it in the config.yaml file. By default, config.yaml file uses adaptive threshold that uses validation/tests sets.

**Reply by ngadhvi** [reply]
Hey @samet-akcay, thank you for explaining but I was wondering how can this manual threshold be implemented? Here is the part of the config.yaml with the threshold part;

threshold:
    image_default: 3
    pixel_default: 3
    adaptive: true      # **should just setting this to false will work?**

**Reply by nicozhou7** [reply]
Is this solution out now? It will be nice to have it

**Reply by djdameln** [reply]
We currently have a proof of concept implemented for this unsupervised thresholding mechanism, which you can find in [this PR](https://github.com/openvinotoolkit/anomalib/pull/634). The solution works by generating synthetic anomalous images from a subset of the normal validation images by augmenting the images with a randomly generated nosie mask. These synthetic anomalous images are included in the validation set during adaptive thresholding. Please note that this solution is still in an experimental state, so we can't guarantee that it will improve results for every model and the code might not be fully stable.

**Reply by nicozhou7** [reply]
oh wow, nice solution, how can I possibly use this PR? do I need to wait until it merged?

**Reply by nicozhou7** [reply]
So anomalib is calculating the threshold using the validation data only right?

**Reply by djdameln** [reply]
Since the feature is still under construction, it would probably be best to wait for one of the next releases which will include this feature.

Yes, Anomalib's adaptive thresholding mechanism relies only on the validation data.

**Reply by nicozhou7** [reply]
Thanks for replying. But i could play with the poc features using your branch right? cause im also curious with the result. And also does not sure when will it be merged and released XD.

**Reply by djdameln** [reply]
Yes, you could definitely check out [the branch](da/synthetic-validation-set). Just be aware that the feature is still under construction and we cannot guarantee that it is stable.

**Reply by nicozhou7** [reply]
Yeah this is what I am looking for !!!! Cause I really want a fully unsupervised method, even for calculating the threshold.

**Comment by samet-akcay** [comment]
@djdameln would you be able to comment here?

**Reply by valroshan** [reply]
While using Anomaly 2.2.0 with patchcore and datamodule method , after i run the engine , i find only some of the images from my test set are participating , what needs to be done for making all the images participate so that the results will become more accurate , can you pl tell me guide me on the flow , right versions , and correct folder structure
siva

**Comment by djdameln** [comment, marked as answer]
Setting the adaptive threshold parameter to `false` should be sufficient to prevent anomalib from computing the optimal threshold value based on the validation set. It will then use the entered values for the `image_default` and `pixel_default` parameters instead. The values of these parameters should be chosen by trial and error, and may vary between models and datasets.

Please note however that even when adaptive thresholding is disabled, Anomalib still expects some abnormal images to be provided. This is because by design, anomalib evaluates each model after training in order to present some performance metrics to the user. This design is based on the assumption that for any use case, there will be at least a few examples available of the anomalous class.

I feel it would be good to remove this constraint, and allow training a model without performing evaluation afterwards. This would at least require some changes to the dataset classes, but possibly other parts of the pipeline will be affected as well.

**Reply by nicozhou7** [reply]
I found it can do validation using only good images to calculate threshold, though the metrics will be not meaningful, is it make senses?

**Reply by djdameln** [reply]
That's correct. When no anomalous images are available, the adaptive thresholding mechanism will not produce a meaningful threshold value. This is because the adaptive threshold is computed as the value that maximizes the F1 score over the validation set, and the F1 score cannot be computed without positive (anomalous) samples in the ground truth dataset.

**Reply by nicozhou7** [reply]
But the thing is that, it could still be used to catch some anomaly. Though not as good as I wanted it to be(lots of overkill). So, does it means the threshold is somehow working?

**Reply by djdameln** [reply]
When the validation set does not contain any anomalous images, the adaptive thresholding mechanism will effectively return the maximum observed pixel value in the anomaly heatmaps of the normal images. You are right, this might actually lead to decent performance. Though it will always be less accurate than an optimized threshold value that is based on both normal and anomalous images.

**Reply by nicozhou7** [reply]
Many thanks for clarification!

**Reply by nicozhou7** [reply]
But may I know which part of the code did this mechanism?

**Reply by djdameln** [reply]
The adaptive threshold is computed by the [AdaptiveThreshold](https://github.com/openvinotoolkit/anomalib/blob/main/anomalib/utils/metrics/anomaly_score_threshold.py) class, which is a torchmetrics [Metric](https://torchmetrics.readthedocs.io/en/stable/pages/implement.html#torchmetrics.Metric) object. The class is called [here](https://github.com/openvinotoolkit/anomalib/blob/main/anomalib/models/components/base/anomaly_module.py#L132) in the base anomaly module at the end of the validation epoch.

**Reply by nicozhou7** [reply]
Thanks, @djdameln  I am sorry but i could not see where does it effectively return the maximum observed pixel value in the anomaly heatmaps of the normal images?

**Reply by djdameln** [reply]
The maximal observed pixel value is not retrieved explicitly, but implicitly as a side effect of the precision-recall curve computation which is a part of the adaptive threshold computation. The precision-recall curve is computed by calculating the precision and recall scores over the validation set for a range of threshold values. The candidate values for the threshold at which precision and recall are evaluated are obtained by incrementally lowering the anomaly score between the highest and lowest value in the predictions, while monitoring for changes in the precision and recall scores. 

If the validation set has no anomalous samples, then the true positive and false negative counts will both be zero for every possible threshold value. This means that precision (tp/(tp+fp)) will always yield zero, and recall (tp/(tp+fn)) cannot be computed due to zero in the denominator, and will therefore also default to zero.

So, both precision and recall will not change when lowering the value of the threshold candidate. As a result, we only retrieve a single threshold candidate, which coincides with the value of the highest score that occurs in the predictions on the validation set
